File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/api/NfcRendezvousClient.kt
// C3 Wave 2 (Android) — cloud-rendezvous deposit client.
//
// Mirrors POST https://flagshipserver.com/api/nfc/rendezvous/:id/wifi
// from packages/control-plane/src/nfcRendezvous.ts.
//
// Body shape: { sealedHex, nonceHex } — both lowercase hex. The Worker
// rejects nonceHex != 24 chars (12-byte AEAD nonce) and caps sealedHex
// at 16384 hex chars (8 KB sealed-blob abuse cap). `sealedHex` is the
// protocol deposit blob: ePhonePub(32) || AEAD ciphertext (see
// `buildWifiDepositBlob` in core.NfcPair — the box needs the pub to
// derive K_session; tampering it in transit fails the AEAD open). The
// phone-side blob is well under 1 KB in practice.

package com.flagshipserver.app.api

import com.flagshipserver.app.core.JsonHttpTransport
import kotlinx.serialization.Serializable

interface NfcRendezvousClient {
    /** Deposit the deposit blob (hex) + AEAD nonce (hex) at the given
     *  rendezvous slot. Returns Result.failure when the Worker rejects
     *  the deposit. */
    suspend fun depositSealedWifi(rendezvousId: String, sealedHex: String, nonceHex: String): Result<Unit>
}

@Serializable
data class NfcRendezvousDepositRequest(
    val sealedHex: String,
    val nonceHex: String,
)

@Serializable
data class NfcRendezvousDepositResponse(
    val ok: Boolean,
    /** Wall-clock ms after which the deposited slot expires. */
    val expiresAt: Long,
)

// ── Live ──────────────────────────────────────────────────────────

class LiveNfcRendezvousClient(
    private val transport: JsonHttpTransport,
    baseUrl: String = DEFAULT_BASE_URL,
) : NfcRendezvousClient {
    private val base = baseUrl.trimEnd('/')

    companion object {
        const val DEFAULT_BASE_URL = "https://flagshipserver.com"
    }

    override suspend fun depositSealedWifi(
        rendezvousId: String,
        sealedHex: String,
        nonceHex: String,
    ): Result<Unit> {
        val encoded = java.net.URLEncoder.encode(rendezvousId, "UTF-8")
        val body = NfcRendezvousDepositRequest(
            sealedHex = sealedHex,
            nonceHex = nonceHex,
        )
        return try {
            transport.postJsonForResponse(
                url = "$base/api/nfc/rendezvous/$encoded/wifi",
                body = body,
                serializer = NfcRendezvousDepositRequest.serializer(),
                responseSerializer = NfcRendezvousDepositResponse.serializer(),
            )
            Result.success(Unit)
        } catch (t: Throwable) {
            Result.failure(t)
        }
    }
}

// ── Mock ──────────────────────────────────────────────────────────

/**
 * Test seam — records every deposit + lets callers script the next
 * outcome (e.g. simulate a 500 from the Worker). The captured deposit
 * round-trips through openWiFiConfig in the happy-path test.
 */
class MockNfcRendezvousClient(
    var nextOutcome: Result<Unit> = Result.success(Unit),
) : NfcRendezvousClient {
    data class Deposit(val rendezvousId: String, val sealedHex: String, val nonceHex: String)

    private val _deposits = mutableListOf<Deposit>()
    val deposits: List<Deposit> get() = _deposits

    override suspend fun depositSealedWifi(
        rendezvousId: String,
        sealedHex: String,
        nonceHex: String,
    ): Result<Unit> {
        _deposits += Deposit(rendezvousId, sealedHex, nonceHex)
        return nextOutcome
    }
}

File: apps/mobile/shared/Sources/FlagshipAPI/Client/NfcRendezvousClient.swift
import Foundation

/// C3 — phone client for the NFC rendezvous drop-box on flagshipserver.com.
///
/// Wire shape (see `packages/control-plane/src/nfcRendezvous.ts`):
///
///   POST /api/nfc/rendezvous/:rendezvousId/wifi
///     Body: { "sealedHex": "<hex>", "nonceHex": "<hex>" }
///     → 200 { ok: true, expiresAt: number }
///     → 400 malformed
///     → 429 rate limited
///
/// `sealedHex` is the protocol deposit blob: the phone's ephemeral
/// X25519 public key (32 bytes) followed by the AES-GCM ciphertext
/// sealed under K_session (`buildWifiDepositBlob` in FlagshipCore —
/// the box needs the pub to derive the same key; tampering it in
/// transit fails the AEAD open). The cloud is a pure opaque relay — anyone who
/// guesses both a live `rendezvousId` AND the blob format can deposit
/// garbage, but the box will reject non-genuine deposits when it tries
/// to open them with its independently-derived K_session.
///
/// 8 KB sealed-payload cap enforced server-side; the WiFi-config blob is
/// well under 1 KB so this is just an abuse limit, not an expected size.

public protocol NfcRendezvousClient: Sendable {
    func depositSealedWifi(rendezvousId: String, sealedHex: String, nonceHex: String) async throws
}

public enum NfcRendezvousError: Error, Equatable, Sendable, LocalizedError {
    case badRequest(String)
    case rateLimited
    case http(status: Int, body: String)
    case transport(String)

    public var errorDescription: String? {
        switch self {
        case .badRequest(let m): return "Rendezvous rejected: \(m)"
        case .rateLimited:        return "Too many tries — wait a moment and try again."
        case .http(let s, let b): return "Server returned \(s): \(b)"
        case .transport(let m):   return "Network error: \(m)"
        }
    }
}

// MARK: - Live

public final class LiveNfcRendezvousClient: NfcRendezvousClient, @unchecked Sendable {
    public static let defaultBaseUrl = URL(string: "https://flagshipserver.com")!

    private let urlSession: URLSession
    private let baseUrl: URL

    public init(urlSession: URLSession = .shared, baseUrl: URL = defaultBaseUrl) {
        self.urlSession = urlSession
        self.baseUrl = baseUrl
    }

    public func depositSealedWifi(
        rendezvousId: String,
        sealedHex: String,
        nonceHex: String
    ) async throws {
        // Server-side regex is `^[A-Za-z0-9_-]{8,64}$`; rejecting client-
        // side too gives a clearer error than a 400 from the edge.
        guard rendezvousId.range(of: "^[A-Za-z0-9_-]{8,64}$", options: .regularExpression) != nil else {
            throw NfcRendezvousError.badRequest("rendezvousId must match [A-Za-z0-9_-]{8,64}")
        }
        let encodedId = rendezvousId.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? rendezvousId
        let url = baseUrl.appendingPathComponent("/api/nfc/rendezvous/\(encodedId)/wifi")
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "content-type")
        req.httpBody = try JSONSerialization.data(withJSONObject: [
            "sealedHex": sealedHex,
            "nonceHex": nonceHex,
        ])

        let data: Data
        let resp: URLResponse
        do {
            (data, resp) = try await urlSession.data(for: req)
        } catch {
            throw NfcRendezvousError.transport(error.localizedDescription)
        }
        let status = (resp as? HTTPURLResponse)?.statusCode ?? 0
        if (200..<300).contains(status) { return }
        let body = String(data: data, encoding: .utf8) ?? ""
        switch status {
        case 400: throw NfcRendezvousError.badRequest(body)
        case 429: throw NfcRendezvousError.rateLimited
        default:  throw NfcRendezvousError.http(status: status, body: body)
        }
    }
}

// MARK: - Mock

/// Scriptable mock for tests. Default behavior = success; set `behavior`
/// to throw a specific error. Captures the last call's args so tests can
/// assert wire shape (URL, hex values, retry count).
public final class MockNfcRendezvousClient: NfcRendezvousClient, @unchecked Sendable {
    public enum Behavior: Sendable {
        case ok
        case failure(NfcRendezvousError)
    }
    public var behavior: Behavior = .ok
    public private(set) var lastDeposit: (rendezvousId: String, sealedHex: String, nonceHex: String)?
    public private(set) var callCount: Int = 0

    public init() {}

    public func depositSealedWifi(
        rendezvousId: String,
        sealedHex: String,
        nonceHex: String
    ) async throws {
        callCount += 1
        lastDeposit = (rendezvousId, sealedHex, nonceHex)
        switch behavior {
        case .ok: return
        case .failure(let e): throw e
        }
    }
}
